ee/hogai/notebook/notebook_serializer.py

```diff
@@ -0,0 +1,610 @@
+            elif pattern_type == "link":
+                link_text = pattern_data["text"]
+                href = pattern_data["href"]
+                if self._is_safe_url(href):
+                    content.append(
+                        ProsemirrorJSONContent(
+                            type="text",
+                            text=link_text,
+                            marks=[Mark(type="link", attrs={"href": href, "target": "_blank"})],
+                        )
+                    )
+                else:
+                    # Unsafe URL, just add as text
+                    content.append(ProsemirrorJSONContent(type="text", text=link_text))
+
+            pos = match_end
+
+        return content if content else [ProsemirrorJSONContent(type="text", text=text)]
+
+    def _find_next_markdown_pattern(self, text: str, start_pos: int) -> Optional[tuple[int, int, str, dict]]:
+        """Find the next markdown formatting pattern in text."""
+        patterns = [
+            # Bold: **text** or __text__ - check these first to prioritize over italic
+            (r"\*\*(.+?)\*\*", "bold"),
+            (r"__(.*?)__", "bold"),
+            # Italic: *text* or _text_
+            (r"\*(.*?)\*", "italic"),
+            (r"_(.*?)_", "italic"),
+            # Code: `text`
+            (r"`(.*?)`", "code"),
+            # Strikethrough: ~~text~~
+            (r"~~(.*?)~~", "strikethrough"),
+            # Link: [text](url)
+            (r"\[([^\]]*)\]\(([^)]*)\)", "link"),
+        ]
+
+        earliest_match = None
+        earliest_pos = len(text)
+
+        for pattern, pattern_type in patterns:
+            match = re.search(pattern, text[start_pos:])
+            if match:
+                match_start = start_pos + match.start()
+                match_end = start_pos + match.end()
+
+                if match_start < earliest_pos:
+                    earliest_pos = match_start
+                    if pattern_type == "link":
+                        earliest_match = (
+                            match_start,
+                            match_end,
+                            pattern_type,
+                            {"text": match.group(1), "href": match.group(2)},
+                        )
+                    else:
+                        earliest_match = (match_start, match_end, pattern_type, {"text": match.group(1)})
+
+        return earliest_match
+
+    def _parse_blockquote_content(self, content: str) -> list[ProsemirrorJSONContent]:
+        """Parse blockquote content as nested markdown."""
+        # Recursively parse the blockquote content as markdown
+        tokenizer = MarkdownTokenizer()
+        tokens = tokenizer.tokenize(content)
+
+        result = []
+        for token in tokens:
+            nodes = self._convert_markdown_token(token)
+            result.extend(nodes)
+
+        return result if result else [self.to_json_paragraph("")]
+
+    def _is_safe_url(self, url: str) -> bool:
```

Comment:
If to be paranoid - it seems it could still be unsafe 😅  Like, if we add something like `%256Aavascript:` it would decode to `%6Aavascript:` if we do just a single unquote, so there's a space for XSS attack, right now something like `'javascript%253Aalert(1)` could pass through.

Could be skipped, but if we want to be double sure it would probably make sense to have a  recursive checj. Not sure how much do we care though, or where the URL can come from.